proposals/lip-0051.md

```diff
@@ -769,14 +770,222 @@ if the above send function does not fail:
+                }
+            }
+        },
+        "availableLocalIDSubstore": {
+            "dataType": "uint32",
+            "fieldNumber": 1
+        },
+        "terminatedEscrowSubstore": {
+            "type": "array",
+            "fieldNumber": 1,
+            "items": {
+                "type": "uint32"
+            }
+        }
+    }
+}
+```
+
+
+#### Genesis State Initialization
+
+During the genesis state initialization stage, the following steps are executed. If any step fails, the block is discarded and has no further effect.
+
+Let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the Token module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisTokenStoreSchema` schema, given above.
+
+* Initial checks:
+   * Accross all elements of the `userSubstore` array, the pairs `address, tokenID` must be unique (each pair appears at most once, though a given address can appear multiple times with different token IDs). 
+   * For each such `address, tokenID` pair, the values used as `moduleID` in the `lockedBalances` array must be unique.
+   * Accross all element of the `supplySubstore` array, all values given for `localID` must be unique.
+   * Accross all elements of the `escrowSubstore` array, the pairs `escrowChainID, localID` must be unique (each pair appears at most once). 
+
+* For each entry `userEntry` in `genesisBlockAssetObject.userSubstore`, create an entry in the user substore  with
+  ```python
+  storeKey = userEntry.address || uint32be(userEntry.tokenID.chainID) || uint32be(userEntry.tokenID.localID)
+  storeValue = {"availableBalance": userEntry.availableBalance,
+                "lockedBalances": [{"moduleID": lockedBalance.moduleID,
+                                    "amount": lockedBalance.amount
+                                   } for each lockedBalance in userEntry.lockedBalances]
+               } serialized using userStoreSchema.
+  ```
+
+* For each entry `supplyEntry` in `genesisBlockAssetObject.supplySubstore`, create an entry in the supply substore  with
+  ```python
+  storeKey = uint32be(supplyEntry.localID)
+  storeValue = {"totalSupply": supplyEntry.totalSupply} serialized using supplyStoreSchema.
+  ```
+
+* For each entry `escrowEntry` in `genesisBlockAssetObject.escrowSubstore`, create an entry in the escrow substore  with
+  ```python
+  storeKey = uint32be(escrowEntry.escrowChainID) || uint32be(escrowEntry.localID)
+  storeValue = {"amount": escrowEntry.amount} serialized using escrowStoreSchema.
+  ```
+
+* Create an entry in the available local ID substore with
+  ```python
+  storeKey = EMPTY_BYTES
+  storeValue = {"nextAvailableLocalID": genesisBlockAssetObject.availableLocalIDSubstore} 
+               serialized using availableLocalIDStoreSchema.
+  ```
+
+* For each entry `terminatedChainID` in `genesisBlockAssetObject.terminatedEscrowSubstore`, create an entry in the terminated escrow substore with
```

Comment:
What purpose would have to start a sidechain with terminated escrows of other chains? 

Also, if there is no corresponding entry in `escrowSubstore` for the terminated chain, would this substore be usable in any way?

Reply:
The purpose is mainly: testing and regenesis. 
As this state can exist, I allowed it to be created at genesis. 
Funnily enough, there is almost no link between chains in the `terminated escrow substore` and tokens in the `escrowSubstore`. The terminated escrow substore only serves as a reminder of chains that have broken the token module rules, regardless of the existence of tokens actually escrowed to that chain.